src/librbd/mirror/snapshot/GroupCreatePrimaryRequest2.h

```diff
@@ -0,0 +1,176 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_LIBRBD_MIRROR_SNAPSHOT_GROUP_CREATE_PRIMARY_REQUEST_H
+#define CEPH_LIBRBD_MIRROR_SNAPSHOT_GROUP_CREATE_PRIMARY_REQUEST_H
+
+#include "include/buffer.h"
+#include "include/rados/librados.hpp"
+#include "librbd/ImageWatcher.h"
+#include "cls/rbd/cls_rbd_types.h"
+#include "librbd/internal.h"
+#include "librbd/mirror/snapshot/Types.h"
+
+#include <string>
+#include <set>
+
+struct Context;
+
+namespace librbd {
+
+struct ImageCtx;
+
+namespace mirror {
+namespace snapshot {
+
+template <typename ImageCtxT = librbd::ImageCtx>
+class GroupCreatePrimaryRequest2 {
```

Comment:
Can you please explain at a high level what `GroupCreatePrimaryRequest2` is different from `GroupCreatePrimaryRequest` state machine?

I see it calls `CreatePrimaryRequest2`.  I think there is more to it than that.

Reply:
These are eventually expected to replace for the current state machines - CreatePrimaryRequest2 will be renamed to CreatePrimaryRequest and all callers of the current CreatePrimaryRequest will switch to using it. For standalone images, the vectors will contain a single element.

GroupCreatePrimary2 just calls  CreatePrimaryRequest2 to which offloads the exclusive locking and quiesce operations. This leads to the locks being held for a shorter interval, hence reducing the effect on I/O.